core/src/main/java/org/infinispan/interceptors/locking/AbstractLockingInterceptor.java

```diff
@@ -103,12 +103,16 @@ public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryComm
 
    // We need this method in here because of putForExternalRead
    protected final Object visitNonTxDataWriteCommand(InvocationContext ctx, DataWriteCommand command) throws Throwable {
-      if (hasSkipLocking(command) || !shouldLockKey(command.getKey())) {
+      Object key;
+      // We don't lock the key if the lock owner is the same instance as the key. This a special case where the lock
+      // owner was specifically set by locking code to control it. The controlling code will then unlock the key
+      // themselves.
+      if (hasSkipLocking(command) || ((key = command.getKey()) == ctx.getLockOwner()) || !shouldLockKey(key)) {
```

Comment:
I had thought about this a while back, forgot who I talked to it about it. The problem as you already mentioned is what do we do in the case of putAll/getAll and others and it became way too hairy. So I went with just a different lock owner as that would work for all.